### DISCUSSION TITLE
How to duplicate data "best"?

### DISCUSSION THREAD

**Discussion by rokdd** [directus/directus#14429] [Archive]
Hello 

I just asked some days ago in discord how it is possible to duplicate items by the best solution.

Want I want to achieve:
I want to realize like a versioning by column. So my plan is to put a field "version" and each time I decide by a manual trigger / call, I duplicate all items with version X. My data will use relationsships so also the other connected items need to duplicated as well. All the new items should get a new version Y.

What I know:
I found out that the "Save as copy" in the UI is already very close, because I can also setup which fields are getting duplicated as well (https://github.com/directus/directus/blob/7d451f8df4383373380b7a3eb39d9dfce3488982/app/src/composables/use-item/use-item.ts#L149). I tried to access this function from extensions but it did not suceed sadly :-(

So what is the best way to do this?
- per flow: I guess impossible
- per extension: how to access collections and items in the extensions by their native functions?
- per API from external?
- any other ideas?

Thanks for your help and ideas..

**Comment by br41nslug** [comment]
In my opinion the only way to duplicate data is to do a fetch, filter out the unique data like primarykey and save that. That is what `saveAsCopy` does in the App but somewhat hidden as the data is already fetched to show it. Depending on what stage you want to implement this you have the following options:
- frontend extensions interface/display/etc or external apps you'll need to use the api to fetch and save
- for api extensions like endpoints/hooks you have access to services (like `ItemsService`) used to talk directly to the database.
- might be possible with flows im just not sure how

**Reply by rokdd** [reply]
Thanks for your help! I am developing currently an API endpoint to duplicate my data. It is hard to do it properly, but some findings for everybody else:
- I used the services of directus, because I always try to be close as possible at the software. It turned out that auto fields for dates can be only set by directly updating with knex in the database (https://github.com/directus/directus/discussions/14748). So when you are keen you can consider to duplicate the data by knex without bothering the item services
- my layout is to have a release field in all collections including the relationships and some system collections. Only so it is possible to clean track what is created in which version. All new items got the default for a specific version.
- I also made a collection where I save which id was copied into which new id to prevent duplicating twice
- I also made some mechanism to automatically configure the roles that the editors can only the draft release. The status field I not touched to have some status when editing the current release
- relationships m2m are only visible in the field-service . In the relation-service it get splitted into M2o and o2M relationships seperate
- some revisions and article fields can be only updated by database directly (else the changes appears when you execute your script). I disabled accountability to not circle in changes and filter by only certain types of events

Hope that explains a bit, I am still in progress, else drop me a message.

**Comment by sevetseh28** [comment]
Hey @rokdd any progress on this? I'd like to implement this without having to do a GET and POST because this will put the cloning logic on every client.
We want to add versioning to the aggregate root via a `/items/<rootCollectionName>/<id>/clone` endpoint or something like that, that would copy all of the related entities to the aggregate root collection and create a new clone (including related entities) in a single HTTP call. 

I was thinking of adding a proxy that would hide this GET and POST logic.

**Reply by rokdd** [reply]
I am still working on a solution for the problem, but it is very customized to our needs. I don't really get your issue regarding the client and post/get requests? My way ended up in a extension with one endpoint for cloning all set up collections and its relations. I guess to clone a specific collection you  have to be very careful / aware of relationships and how to handle them. For example when my code was not running very well at the beginning, I accidentally changed the "source" data because I assigned the destination data, items of the source items.
